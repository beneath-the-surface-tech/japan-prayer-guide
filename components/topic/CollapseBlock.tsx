import React, { ReactNode } from "react"
import { Collapse, Container } from "react-bootstrap"
import Button from "react-bootstrap/Button"
import { RiArrowDownSLine, RiArrowUpSLine } from "react-icons/ri"

interface feedbackProps {
    title: string
    children: ReactNode
    startOpened?: boolean
}

const CollapseBlock = ({ title, children, startOpened = false }: feedbackProps) => {
    const [openCollapse, setOpenCollapse] = React.useState(startOpened)

    return (
        <Container>
            <hr />
            <Button
                className="collapse-label fs-2 fw-semibold lh-sm ps-0"
                onClick={() => setOpenCollapse(!openCollapse)}
                aria-controls="collapse-block"
                aria-expanded={openCollapse}
            >
                {title} {openCollapse ? <RiArrowUpSLine /> : <RiArrowDownSLine />}
            </Button>
            <Collapse in={openCollapse}>
                <Container id="collapse-block">{children}</Container>
            </Collapse>
        </Container>
    )
}

export default CollapseBlock
