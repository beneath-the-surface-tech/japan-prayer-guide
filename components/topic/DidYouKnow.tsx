import React from "react";
import { Container } from "react-bootstrap";
import { IconContext } from "react-icons";
import { BsLightbulb } from "react-icons/bs";

interface dykProps {
    text: string
    title: string
}

const DidYouKnow = ({text, title}:dykProps) => {
    return (
        <Container data-testid={"dyk-container"} className="d-flex justify-content-center">
            <div className="did-you-know p-4 shadow">
                <div className="d-flex align-items-center">
                    <IconContext.Provider value={{size: '25px'}}>
                        <BsLightbulb/>
                    </IconContext.Provider>
                    <h1 className="fw-bold fs-5 my-0 mx-2">{title}</h1>
                </div>
                <div data-testid={"dyk-text"} className="fs-5 my-3">
                    {text}
                </div>
            </div>
        </Container>
    )
}

export default DidYouKnow