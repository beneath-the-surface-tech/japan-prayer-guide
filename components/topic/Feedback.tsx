import { Trans, useTranslation } from "next-i18next"
import React from "react"
import Button from "react-bootstrap/Button"

const Feedback = () => {
    const { t } = useTranslation("common")
    return (
        <section
            data-testid={"feedback-sect"}
            className="text-center m-0 p-4 w-100"
            style={{ backgroundColor: "#EAF4FF" }}
        >
            <h1 data-testid={"feedback-title"} className="text-primary fs-2">
                <Trans t={t} i18nKey="feedback.title" />
            </h1>
            <Button className="bg-secondary text-white border-0 p-2 w-100 mt-4" style={{ maxWidth: "300px" }}>
                <p data-testid={"feedback-button"} className="m-0">
                    <Trans t={t} i18nKey="feedback.button" />
                </p>
            </Button>
        </section>
    )
}

export default Feedback
