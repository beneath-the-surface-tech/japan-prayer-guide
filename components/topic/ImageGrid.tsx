import React from "react";
import { Container, Carousel } from "react-bootstrap";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

const ImageGroup = () => {
    return (
        <Container className="w-100 d-flex justify-content-center">
            <ImageGrid/>
            <ImageCarousel/>
        </Container>
    )
}

const ImageGrid = () => {
    return (
        <div className="w-100 d-flex flex-column gap-3 d-none d-lg-flex my-3">
            <Row className="">
                <Col>
                    <div id="about-landing-image" className=""></div>
                </Col>
                <Col>
                    <div id="about-landing-image" className=""></div>
                </Col>
            </Row>
            <Row>
                <Col>
                    <div id="about-landing-image" className=""></div>
                </Col>
                <Col>
                    <div id="about-landing-image" className=""></div>
                </Col>
            </Row>
        </div>
    )
}

const ImageCarousel = () => {
    return (
        <Carousel id="image-carousel" className="w-100 d-lg-none" interval={null}>
            <Carousel.Item>
                <div id="about-landing-image" className=""></div>
            </Carousel.Item>
            <Carousel.Item>
                <div id="about-landing-image" className=""></div>
            </Carousel.Item>
            <Carousel.Item>
                <div id="about-landing-image" className=""></div>
            </Carousel.Item>
            <Carousel.Item>
                <div id="about-landing-image" className=""></div>
            </Carousel.Item>
        </Carousel>
    )
}

export default ImageGroup