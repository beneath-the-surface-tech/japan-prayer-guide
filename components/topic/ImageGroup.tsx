import React from "react"
import { Container, Carousel } from "react-bootstrap"
import Row from "react-bootstrap/Row"
import Col from "react-bootstrap/Col"

// TODO: Should accept an array of image links which are displayed here

const ImageGroup = () => {
  return (
    <Container className="w-100 d-flex justify-content-center">
      <ImageGrid />
      <ImageCarousel />
    </Container>
  )
}

const ImageGrid = () => {
  return (
    <div className="w-100 d-flex flex-column gap-3 d-none d-md-flex my-3">
      <Row>
        <Col>
          <div className="placeholder-image"></div>
        </Col>
        <Col>
          <div className="placeholder-image"></div>
        </Col>
      </Row>
      <Row>
        <Col>
          <div className="placeholder-image"></div>
        </Col>
        <Col>
          <div className="placeholder-image"></div>
        </Col>
      </Row>
    </div>
  )
}

const ImageCarousel = () => {
  return (
    <Carousel id="image-carousel" className="w-100 d-md-none" interval={null}>
      <Carousel.Item>
        <div className="placeholder-image"></div>
      </Carousel.Item>
      <Carousel.Item>
        <div className="placeholder-image"></div>
      </Carousel.Item>
      <Carousel.Item>
        <div className="placeholder-image"></div>
      </Carousel.Item>
      <Carousel.Item>
        <div className="placeholder-image"></div>
      </Carousel.Item>
    </Carousel>
  )
}

export default ImageGroup
