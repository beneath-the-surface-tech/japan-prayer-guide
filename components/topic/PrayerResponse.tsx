import { TFunction, Trans, useTranslation } from "next-i18next"
import React from "react"
import { Col, Container, FloatingLabel, Form, Row } from "react-bootstrap"
import Button from "react-bootstrap/Button"
import PrayerPoints, { PrayerDisplayStyle } from "../common/PrayerPoints"

const PrayerResponse = ({ topicTrans }: { topicTrans: TFunction }) => {
    const { t: common } = useTranslation("common")

    const textPlaceholder: string = common("response.feedbackTextPlaceholder", "")
    const emailPlaceholder: string = common("response.feedbackEmailPlaceholder", "")

    const topicVideoSrc: string = topicTrans("videoSrc", "")
    const couldLoadVideo: boolean = topicVideoSrc !== ""

    return (
        <Container fluid id="prayer-response" data-testid="prayer-response" className="pb-5">
            <Container>
                <h2 className="fw-bold text-primary">
                    <Trans t={common} i18nKey="response.title" />
                </h2>
                <p>
                    <Trans t={common} i18nKey="response.blurb0" />
                </p>
                <p>
                    <Trans t={common} i18nKey="response.blurb1" />
                </p>

                <h3 className="fw-bold fs-2">
                    <Trans t={common} i18nKey="response.recordingLabel" />
                </h3>
                <div className="mb-5 video-embed">
                    {couldLoadVideo ? (
                        <iframe
                            className="d-flex align-content-start"
                            src={topicVideoSrc}
                            allow="autoplay; fullscreen"
                            style={{
                                position: "absolute",
                                top: "0",
                                left: "0",
                                width: "100%",
                                height: "100%",
                                border: "0px",
                            }}
                            title="Prayer recording"
                        />
                    ) : (
                        <div>
                            <Trans t={common} i18nKey="errors.videoLoadError" />
                        </div>
                    )}
                </div>
                <Row>
                    <Col lg={6} md={12} className="px-0">
                        <PrayerPoints topicTrans={topicTrans} displayStyle={PrayerDisplayStyle.TopicBottom} />
                    </Col>
                    <Col lg={6} md={12}>
                        <Container data-testid="feedback" fluid className="px-0">
                            <h3 className="fw-bold mt-0">
                                <Trans t={common} i18nKey="response.feedbackLabel" />
                            </h3>
                            <Form action="https://formbold.com/s/67Gj1" method="POST">
                                <Form.Group className="mb-2">
                                    <FloatingLabel
                                        controlId="floatingTextArea"
                                        label={textPlaceholder}
                                        className="text-muted"
                                    >
                                        <Form.Control
                                            as="textarea"
                                            name="message"
                                            placeholder={textPlaceholder}
                                            style={{ height: "100px" }}
                                        />
                                    </FloatingLabel>
                                </Form.Group>
                                <Form.Group className="mb-3">
                                    <Form.Control type="email" placeholder={emailPlaceholder} />
                                </Form.Group>
                                <Button variant="outline-primary" className="fw-bold px-4 py-2" type="submit">
                                    <Trans t={common} i18nKey="response.feedbackSubmit" />
                                </Button>
                            </Form>
                        </Container>
                    </Col>
                </Row>
            </Container>
        </Container>
    )
}

export default PrayerResponse
