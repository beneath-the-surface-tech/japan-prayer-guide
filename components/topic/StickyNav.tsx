import { Trans } from "next-i18next"
import { useEffect, useRef, useState } from "react"
import { Container } from "react-bootstrap"

const scrollOffset = -50

export interface Tab {
    refId: string
    label: string
}

interface StickyNavProps {
    tabs: Tab[]
}

export const StickyNav = ({ tabs }: StickyNavProps) => {
    const bannerRef = useRef<HTMLDivElement>(null)
    const [selected, setSelected] = useState<string>(tabs[0].refId)
    const [show, setShow] = useState(false)
    const isScrolling = useRef(false)

    const highlightCurrentNavLink = () => {
        const scrollY = window.scrollY
        // The leeway makes it so that if they click the link and scroll a little, it will still show the correct section as selected
        const scrollLeeway = 10

        const currentScrolledSection = tabs
            .map((tab) => {
                const section = document.getElementById(tab.refId)
                const tabWithPosition = {
                    tab: tab,
                    sectionTop: -1,
                }
                if (section) {
                    tabWithPosition.sectionTop =
                        section.getBoundingClientRect().top + window.pageYOffset + 2 * scrollOffset - scrollLeeway
                }

                return tabWithPosition
            })
            .sort((a, b) => a.sectionTop - b.sectionTop) // sort ascending
            .filter((tab) => tab.sectionTop > 0 && scrollY >= tab.sectionTop) // remove invalid sections and elements not yet scrolled to
            .pop()
        if (currentScrolledSection === undefined) {
            // in the case where we have yet to scroll to any of the sections, set the first section as selected
            setSelected(tabs[0].refId)
        } else {
            setSelected(currentScrolledSection.tab.refId)
        }
    }

    useEffect(() => {
        const handleScroll = () => {
            const bannerElement: any = bannerRef.current
            if (bannerElement) {
                const { top } = bannerElement.getBoundingClientRect()
                if (top === 0) {
                    setShow(true)
                } else {
                    setShow(false)
                }
            }

            if (!isScrolling.current) {
                highlightCurrentNavLink()
            }
        }

        window.addEventListener("scroll", handleScroll)
        return () => {
            window.removeEventListener("scroll", handleScroll)
        }
    }, [])

    const scrollToCustomPosition = (id: string) => {
        setSelected(id)
        isScrolling.current = true // block the listener from triggering while we scroll
        const targetElement = document.getElementById(id)

        if (targetElement) {
            const topPosition = targetElement.getBoundingClientRect().top
            const offset = 2 * scrollOffset // main nav + sticky nav
            window.scrollTo({ top: topPosition + window.pageYOffset + offset, behavior: "smooth" })

            setTimeout(() => {
                isScrolling.current = false
            }, 1500)
        }
    }

    return (
        <Container
            style={{ maxWidth: "none" }}
            ref={bannerRef}
            data-testid={"sticky-nav-container"}
            className="w-100 d-flex align-items-center justify-content-center flex-column sticky-top p-0 sticky-nav"
        >
            <div id="overview-nav" className="w-100" style={show ? {} : { height: "0px" }}></div>
            <div className="w-100 d-flex justify-content-center bg-white">
                <div className="limit-width d-flex justify-content-center border-bottom">
                    <div
                        data-testid={"topic-nav-links"}
                        id="topic-nav-links"
                        className="pt-4 pb-0 w-100 border-bottom d-flex justify-content-center bg-white text-center"
                    >
                        {tabs.map((tab, idx) => (
                            <a
                                key={tab.refId}
                                className={
                                    "px-3 my-0 text-decoration-none fw-normal w-100" +
                                    (selected === tab.refId
                                        ? " border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
                                        : "")
                                }
                                onClick={() => scrollToCustomPosition(tab.refId)}
                                data-testid={"topic-nav-link" + idx}
                            >
                                <Trans>{tab.label}</Trans>
                            </a>
                        ))}
                    </div>
                </div>
            </div>
        </Container>
    )
}
