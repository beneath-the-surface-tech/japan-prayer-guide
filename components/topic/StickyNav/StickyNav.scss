.sticky-nav {
    position: -webkit-sticky;
    padding-top: 50px;
    margin-top: -50px;
    @include media-breakpoint-down(md) {
        margin-top: -56px;
    }
    a {
        font-size: 1.125rem;
        color: black;
        cursor: pointer;
    }
}
