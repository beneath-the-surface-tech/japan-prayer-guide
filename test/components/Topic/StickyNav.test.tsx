import { fireEvent, render, screen } from "@testing-library/react"
import { StickyNav, Tab } from "../../../components/topic/StickyNav"

jest.mock("next-i18next", () => ({
    // this mock makes sure any components using the translate hook can use it without a warning being shown
    useTranslation: () => {
        return {
            t: (str: any) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        }
    },
    Trans: (props: any) => {
        if (props.children) {
            // mock for <Trans>{value}</Trans>
            return props.children
        } else if (props.i18nKey) {
            // mock for <Trans t={t} i18nKey={value} />
            return props.i18nKey
        }
        return props
    },
}))

describe("Sticky Nav", () => {
    test("Snapshot test", () => {
        // This test is to check against unintended changes.
        // If the change is intentional you can update the snapshot with `jest --updateSnapshot`

        const tabs: Tab[] = [
            { refId: "a", label: "abc" },
            { refId: "b", label: "def" },
            { refId: "c", label: "ghi" },
        ]

        const component = render(<StickyNav tabs={tabs} />)
        expect(component).toMatchSnapshot()
    })

    test("Renders sticky-top nav section with the right labels", () => {
        const tabs: Tab[] = [
            { refId: "about-id", label: "about" },
            { refId: "prayer-id", label: "prayer" },
            { refId: "resources-id", label: "resources" },
        ]

        render(<StickyNav tabs={tabs} />)
        const navCont = screen.getByTestId("sticky-nav-container")
        const navLinks = screen.getByTestId("topic-nav-links")
        const nav0 = screen.getByTestId("topic-nav-link0")
        const nav1 = screen.getByTestId("topic-nav-link1")
        const nav2 = screen.getByTestId("topic-nav-link2")

        expect(navCont).toHaveClass("sticky-top", "d-flex", "container", "justify-content-center", "flex-column")

        expect(navLinks).toHaveClass("pt-4", "pb-0", "w-100")
        expect(navLinks.children.length).toBe(3)

        expect(nav0).toHaveTextContent("about")
        expect(nav0).toHaveClass("px-3", "my-0", "text-decoration-none", "fw-normal")
        expect(nav1).toHaveTextContent("prayer")
        expect(nav2).toHaveTextContent("resource")
    })

    test("Can render sticky-top nav section with less sections", () => {
        const tabs: Tab[] = [
            { refId: "culture-id", label: "culture" },
            { refId: "church-id", label: "church" },
        ]

        render(<StickyNav tabs={tabs} />)
        const navLinks = screen.getByTestId("topic-nav-links")
        const nav0 = screen.getByTestId("topic-nav-link0")
        const nav1 = screen.getByTestId("topic-nav-link1")

        expect(navLinks).toHaveClass("pt-4", "pb-0", "w-100")
        expect(navLinks.children.length).toBe(2)

        expect(nav0).toHaveTextContent("culture")
        expect(nav1).toHaveTextContent("church")
    })
})

const TestComponent = () => {
    const tabs: Tab[] = [
        { refId: "culture", label: "Culture and society" },
        { refId: "church", label: "Church and mission" },
    ]

    let mockTop = 500
    HTMLElement.prototype.getBoundingClientRect = jest.fn(() => ({
        width: 10,
        height: 10,
        top: mockTop++,
        left: 0,
        right: 10,
        bottom: 10,
        x: 0,
        y: 0,
        toJSON: jest.fn(),
    }))

    return (
        <>
            <StickyNav tabs={tabs} />
            <div data-testid="abc" id="culture" className="ack"></div>
            <div data-testid="def" id="church"></div>
        </>
    )
}

describe("StickyNav - scrolling behavior", () => {
    test("scrolling to a section activates the relevant nav link", () => {
        render(<TestComponent />)

        const topNavLink = screen.getByTestId("topic-nav-link0")
        const botNavLink = screen.getByTestId("topic-nav-link1")
        expect(topNavLink).toHaveClass("fw-bold", "text-secondary-5")
        expect(botNavLink).not.toHaveClass("fw-bold", "text-secondary-5")

        // Simulate scrolling to the bottom 'church' section, note this number only corresponds to the test
        window.scrollY = 508
        fireEvent.scroll(window)

        expect(topNavLink).not.toHaveClass("fw-bold", "text-secondary-5")
        expect(botNavLink).toHaveClass("fw-bold", "text-secondary-5")
    })
})
