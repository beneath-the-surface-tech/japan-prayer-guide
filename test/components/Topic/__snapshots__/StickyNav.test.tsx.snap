// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sticky Nav Snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-100 d-flex align-items-center justify-content-center flex-column sticky-top px-0 sticky-nav container"
        data-testid="sticky-nav-container"
        style="max-width: none;"
      >
        <div
          class="w-100"
          id="overview-nav"
          style="height: 0px;"
        />
        <div
          class="pt-4 pb-0 w-100 border-bottom d-flex justify-content-center bg-white"
          data-testid="topic-nav-links"
        >
          <a
            class="px-3 my-0 text-decoration-none fw-normal border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
            data-testid="topic-nav-link0"
          >
            abc
          </a>
          <a
            class="px-3 my-0 text-decoration-none fw-normal"
            data-testid="topic-nav-link1"
          >
            def
          </a>
          <a
            class="px-3 my-0 text-decoration-none fw-normal"
            data-testid="topic-nav-link2"
          >
            ghi
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-100 d-flex align-items-center justify-content-center flex-column sticky-top px-0 sticky-nav container"
      data-testid="sticky-nav-container"
      style="max-width: none;"
    >
      <div
        class="w-100"
        id="overview-nav"
        style="height: 0px;"
      />
      <div
        class="pt-4 pb-0 w-100 border-bottom d-flex justify-content-center bg-white"
        data-testid="topic-nav-links"
      >
        <a
          class="px-3 my-0 text-decoration-none fw-normal border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
          data-testid="topic-nav-link0"
        >
          abc
        </a>
        <a
          class="px-3 my-0 text-decoration-none fw-normal"
          data-testid="topic-nav-link1"
        >
          def
        </a>
        <a
          class="px-3 my-0 text-decoration-none fw-normal"
          data-testid="topic-nav-link2"
        >
          ghi
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
